src/simplify_algebra.cpp

```diff
@@ -78,6 +77,162 @@ struct find_mul_conv
+            });
+        if(not std::all_of(is_multi.begin(), is_multi.end(), [](auto b) { return b; }))
+        {
+            return;
+        }
+
+        // output must be multiple of inputs
+        std::vector<std::size_t> scales(in_lens.size());
+        std::transform(
+            in_lens.begin(), in_lens.end(), out_lens.begin(), scales.begin(), [](auto x, auto y) {
+                return y / x;
+            });
+        if(std::any_of(scales.begin(), scales.end(), [](auto i) { return i == 0; }))
+        {
+            return;
+        }
+
+        // wrap up shapes for multibroadcast
+        std::vector<int64_t> in_dims(in_lens.begin(), in_lens.end());
+        std::vector<int64_t> out_dims(out_lens.begin(), out_lens.end());
+        for(int ii = static_cast<int>(scales.size()) - 1; ii >= 0; --ii)
+        {
+            if(scales[ii] == 1 or in_dims[ii] == 1)
+            {
+                continue;
+            }
+
+            out_dims[ii] = in_dims[ii];
+            in_dims.insert(in_dims.begin() + ii + 1, 1);
+            out_dims.insert(out_dims.begin() + ii + 1, scales[ii]);
+        }
+
+        auto in_rsp   = ins_rsp->inputs().front();
+        auto rsp_data = p.insert_instruction(
+            ins_rsp, migraphx::make_op("reshape", {{"dims", in_dims}}), in_rsp);
+        auto mb_rsp = p.insert_instruction(
+            ins_rsp, migraphx::make_op("multibroadcast", {{"output_lens", out_dims}}), rsp_data);
+
+        // reshape other inputs of the pointwise operator
+        std::vector<instruction_ref> pw_inputs;
+        for(const auto& in : ins_pw->inputs())
+        {
+            if(in == ins)
+            {
+                pw_inputs.push_back(mb_rsp);
+            }
+            else
+            {
+                auto mb_in = p.insert_instruction(
+                    ins_rsp, migraphx::make_op("reshape", {{"dims", out_dims}}), in);
+                pw_inputs.push_back(mb_in);
+            }
+        }
+
+        auto pw_out = p.insert_instruction(ins_pw, ins_pw->get_operator(), pw_inputs);
+        std::vector<int64_t> out_dims_rsp(out_lens.begin(), out_lens.end());
+        p.replace_instruction(
+            ins_pw, migraphx::make_op("reshape", {{"dims", out_dims_rsp}}), pw_out);
+    }
+};
+
+struct find_where
+{
+    auto matcher() const
+    {
+        return match::name("gather")(match::args(match::name("concat")(match::arg(0).bind("data")),
+                                                 match::is_constant().bind("ind")));
+    }
+
+    void apply(module& p, match::matcher_result r) const
+    {
+        auto ins      = r.result;
+        auto ins_data = r.instructions["data"];
+        auto ins_ind  = r.instructions["ind"];
+
+        // if ind is not constant, cannot optimize
+        std::vector<bool> vec_ind;
+        auto arg_ind = ins_ind->eval();
+        if(arg_ind.empty())
+        {
+            return;
+        }
+        arg_ind.visit([&](auto v) { vec_ind.assign(v.begin(), v.end()); });
```

Comment:
I think you need to check the values of the indices here as well.

Reply:
added the check